### DISCUSSION TITLE
REPL

### DISCUSSION THREAD

**Discussion by borkdude** [clavascript/clavascript#195] [Ideas]
We need to support a REPL with clavascript and the concept of a development mode.
During development, namespaces need to be fully mutable, so vars can be redefined.
We could focus on a console REPL first, then add nREPL support.

**Comment by lilactown** [comment]
Thinking while typing... can we guarantee a clava REPL we design will work in every environment? Said another way: evaluating code seems inherently coupled with the JS runtime & build tools in use for a given project. Perhaps we should start with a few specific targets... Vite REPL, node.js REPL, deno or bun REPL?

I think that continuing to explore how we could make modules mutable is worthwhile, as I'm not sure anyone has gone deep on that solution in practice. As a second option, we could explore more intimately coupled solutions to the specific environments that we are targeting.

For instance, in the browser using a build tool like Vite, we could potentially use their [hot module reloading API](https://vitejs.dev/guide/api-hmr.html). Most build tools have their own HMR API, so building a plugin for Vite probably won't work for webpack or others.

Curious what you think @borkdude @corasaurus-hex

**Reply by borkdude** [reply]
I think for REPL / dev Clojure vars should just be mapped on a mutable global map, like regular CLJS does it. There is probably no escaping this. For productions builds this can be elided and everything can align nicely with ES modules and is good for tree-shaking.

**Reply by borkdude** [reply]
Honestly, for front-end things like Vite etc. their hot-reloading tooling probably works good enough to not have a REPL. I have to admit that I haven't used a REPL in years for front-end ClojureScript because it was so cumbersome to work with in my experience.

**Reply by stevemolitor** [reply]
I see I can run `squint repl` - awesome! Can I run a nrepl and jack via Cider?

Maybe not so useful for frontend work, but could be super-helpful for node.js scripts.

**Reply by borkdude** [reply]
nREPL not yet. Also this the `squint repl` is very experimental. But you could try to use it with `inf-clojure`.

**Reply by Ralii** [reply]
I personally use REPL quite a lot on frontend. I have emacs functions to inspect app-db, get value from subscriptions, render evaluated component and I use Snitch https://github.com/AbhinavOmprakash/snitch extensively to do interactive frontend development. I understand REPL is not that used in frontend but I really like it is giving me a huge boost in productivity and motivation.

**Reply by borkdude** [reply]
nREPL for Node.js is now available (immature, but please test it and report issues). I hope to enable this on the front-end also somehow.

**Reply by fundchan** [reply]
Injecting REPLs into the runtime is particularly useful for AI development because AI can obtain runtime context information through REPLs.

**Comment by rrrnld** [comment]
Is there any news on nREPL support in the browser? I agree with https://github.com/squint-cljs/squint/discussions/195#discussioncomment-7606772, having a decent REPL in the browser is for me on of the biggest selling points of CLJS. I've dabbled in Squint a bit and I think it's a really nice experiment, but I'm waiting for the REPL support to be mature enough so that I could `cider-jack-in` and evaluate squint forms directly before I use it to build something more serious (can't wait to use reagami for web components!). Is this anywhere on the horizon?

**Reply by borkdude** [reply]
No news yet, but still planning to work on it
